importlib_metadata/__init__.py

```diff
@@ -159,36 +188,40 @@ class EntryPoint(
 
     dist: Optional['Distribution'] = None
 
-    def load(self):
+    def load(self) -> Any:
         """Load the entry point from its definition. If only a module
         is indicated by the value, return that module. Otherwise,
         return the named object.
         """
         match = self.pattern.match(self.value)
+        assert match is not None
```

Comment:
Agreed. I don't expect consumers to be relying on a specific exception. I'm less concerned with changing the exception than I am about the exception changing depending on the optimization flags. One of the features I like about Python is that you don't have to double-check every conditional at run-time. I like writing the minimal amount of code to achieve the primary purpose and hope that exceptions clearly indicate the proximate cause of exceptional conditions. The existing implementation does exactly what I'd want - it _assumes_ the type is a match and fails otherwise. In this case, the assert adds noise to the implementation and runtime variability in the interface.

I don't like how `mypy` is making us second-guess ourselves here. This issue occurs in other places in the code where the need for an assertion causes the need to introduce new scopes of execution (replacing an expression with multiple lines of execution adding names to the local namespace).

Probably the right thing to do here is for Python's regex object to grow a new method (e.g. `ensure_match` or similar) that unconditionally returns a match or raises an exception if no match is found.

Lacking that functionality, perhaps this project should supply that functionality (provide a wrapper that ensures that interface). Such a wrapper could also raise an exception with extra context about what didn't match on what input.

Alternatively, this assertion could be couched in an `if typechecking` block so it's clear that it's typing noise and not essential to the implementation.

```suggestion
        if typing.TYPE_CHECKING:
            assert match is not None
```

On further consideration, I don't like that approach because it still requires breaking out a simple expression into a block of code. I'd say that any solution should be implemented as a wrapper function that's usable in an expression.

```suggestion
        assert_not_none(match)
```

```suggestion
        cast(match, MatchObject)
```